### ISSUE TITLE
[Enhancement]: hot reload Kafka on changes in `brokerCertChainAndKey` instead of a rolling update

### ISSUE THREAD

**Issue by vpedosyuk** [strimzi/strimzi-kafka-operator#9994]
### Related problem

From the [docs](https://strimzi.io/docs/operators/latest/full/configuring.html#property-listener-config-brokerCertChainAndKey-reference):

> When the certificate or key in the `brokerCertChainAndKey` secret is updated, the operator will automatically detect it in the next reconciliation and trigger a rolling update of the Kafka brokers to reload the certificate.

In an environment where a Kafka broker restart is very undesirable, it becomes hard to keep external TLS certificates short-lived (e.g. 24 hours with a 3rd-party PKI) because each change of certificates will cause a Kafka restart and usually a downtime.

In general, it'd be great to have as few reasons for a broker restart as possible.

### Suggested solution

Once a Kubernetes secret referenced in `brokerCertChainAndKey` got changed, Strimzi Operator will dynamically replace old certificates with the new ones without restarting the brokers.

### Alternatives

A proper HA configuration might reduce the effects of such restarts but it's not always possible.

### Additional context

It seems like Kafka itself supports hot-swapping of certificates.

**Comment by scholzj** [comment]
Isn't this already tracked in some other issue? In any case, it should be kept in mind that:
* Improved support for reloading certificates without any major limitations such as DN changes was added only in Kafka 3.7.0. So it is not easy to implement this while supporting Kafka 3.6.x. 
* Updating the configuration is only one part of the problem as you also need to be able to prepare / load the certificates on the fly.
* In reality, the benefits will be also limited as there are parts where TLS certificates might not be reloadable even with Kafka 3.7.0 (for example because they are part of a plugin configurations). So in many setups, the rolling updates will be still needed.

_I do not want to make it sound like this is not worth the effort -> just pointing out that this is not as simple as it might sound and has some obstacles. (I actually wrote the KIP-978 in Kafka exactly for this purpose, it just takes a long time to bubble through)_

**Comment by vpedosyuk** [comment]
@scholzj yes, I've seen your KIP, thanks. In our case SAN and DN remain unchanged, the only thing that changes is expiration time, which is a common case for certificates renewal I believe.

**Comment by scholzj** [comment]
The problem is that unless you can change it all the time, it is basically not feasible because of the complexity. So that is why  that KIP is important as it should allow to use it all the time (for the Kafka parts at least).

**Comment by vpedosyuk** [comment]
Understood. Anyways, thank you for your efforts!

P.S. I couldn't find a similar issue reported here, hence, created one.

**Comment by applejag** [comment]
> - Improved support for reloading certificates without any major limitations such as DN changes was added only in Kafka 3.7.0. So it is not easy to implement this while supporting Kafka 3.6.x.

Now with the release of Strimzi v0.43.0, we don't need to consider Kafka v3.6.x support anymore. At least one thing less to worry about.

**Comment by bh-tt** [comment]
I just ran into this issue, and while this would still be a good idea, the [maintenance window feature](https://strimzi.io/docs/operators/0.45.0/deploying#con-maintenance-time-window-definition-str) for kafka also works quite well for keeping the downtime period contained to an acceptable period. This will not help you much if the certificates change every day though.

**Comment by qlimenoque** [comment]
@scholzj Hi! I'm also waiting for this feature! Do you have any updates on it? :)

**Comment by Neko-Follower** [comment]
Hi! 👋 Any updates on this enhancement?

I’m particularly interested in the ability to hot-reload Kafka when brokerCertChainAndKey changes, instead of triggering a rolling restart. In environments where broker restarts are undesirable, it’s challenging to use short-lived external TLS certificates, since each certificate change currently causes downtime.

Is there any progress or plan for supporting this feature?

Thanks for your hard work on Strimzi!

**Comment by Sorokin41** [comment]
I'm also run into this issue and would really appreciate support for hot-reloading brokerCertChainAndKey. It would be great to know if there’s any progress or plans around making this more seamless!

**Comment by scholzj** [comment]
I'm not aware of anyone working on it.

From my recollection and perspective ... unfortunately, there are still some blockers for dynamically reloading the certificates. Basically, the second and third points from my initial comment are still unsolved. With regards to the second one, it is hard to track the currently used certificates as Kafka treats the fields as Secrets. And because we do not want to reload the certificates again and again, we are kind of stuck on that. So, not sure you should expect progress anytime soon. The plugins are still an issue as well, but that would not affect the listener certificates (although I guess it remains to be seen how feasible it would be to implement it only for them).
